llvm/lib/Target/AMDGPU/SIPreEmitPeephole.cpp

```diff
@@ -326,23 +377,22 @@ bool SIPreEmitPeephole::mustRetainExeczBranch(
       if (TII->hasUnwantedEffectsWhenEXECEmpty(MI))
         return true;
 
-      // These instructions are potentially expensive even if EXEC = 0.
-      if (TII->isSMRD(MI) || TII->isVMEM(MI) || TII->isFLAT(MI) ||
-          TII->isDS(MI) || TII->isWaitcnt(MI.getOpcode()))
-        return true;
-
-      ++NumInstr;
-      if (NumInstr >= SkipThreshold)
+      if (!CostModel.isProfitable(MI))
         return true;
     }
   }
 
   return false;
 }
+} // namespace
 
 // Returns true if the skip branch instruction is removed.
 bool SIPreEmitPeephole::removeExeczBranch(MachineInstr &MI,
                                           MachineBasicBlock &SrcMBB) {
+
+  if (!TII->getSchedModel().hasInstrSchedModelOrItineraries())
```

Comment:
Why do we have a second copy of SchedModel in TII? I thought the subtarget owned it?

Reply:
It seems it's owned by the TII. The `TargetSchedModel` only needs the subtarget for initialization, so it would make sense for it to be a member of the subtarget.

We can move it to the `GCNSubtarget` (it would still be accessible to the TII through the subtarget to be used in `SIInstrInfo::getInstrLatency`).

Additionally, some passes are allocating their own `TargetSchedModel` while they could reference the one allocated by the TII: `GCNHazardRecognizer` and `AMDGPUInsertDelayAlu`. I fixed this in https://github.com/llvm/llvm-project/pull/110859

Reply:
I mean it's already there. I don't see other targets making other copies of this

Reply:
I just realized it was in the base class. I'll fix that such all our references use that one.